### PR TITLE
This adds multiline text widget support.

### LINKED ISSUE #8028: Support specifying system keyboard with carriage return for multiline text editing
Right now the only keyboard action is `TextInputAction.done`, but the ability to set this as a carriage return is need for multiline text editing.

see:
https://developer.android.com/reference/android/text/InputType.html#TYPE_TEXT_FLAG_MULTI_LIN
E

### PR DESCRIPTION
**Pull request by gspencergoog**
This adds support for multi-line text widgets, making a new TextInputAction.newline enum, and supporting a new TextInputType.multiline enum value.

There are still some issues with cursor placement in multi-line input, but those may go away when we switch to a non-Blink text renderer.

Fixes #8028

### CODE REVIEW

File: packages/flutter/lib/src/services/text_input.dart [hunk 1]
```diff
@@ -56,6 +63,9 @@ enum TextInputType {
 enum TextInputAction {
   /// Complete the text input operation.
   done,
+  /// The action to take when the enter button is pressed in a multi-line
```
Comment (Hixie) [inline packages/flutter/lib/src/services/text_input.dart]: nit: missing newline

File: packages/flutter/lib/src/services/text_input.dart [hunk 2]
```diff
@@ -93,13 +106,17 @@ class TextInputConfiguration {
   /// What text to display in the text input control's action button.
   final String actionLabel;
 
+  /// What kind of action to request for the action button on the IME.
+  final TextInputAction inputAction;
+
   /// Returns a representation of this object as a JSON object.
   Map<String, dynamic> toJSON() {
```
Comment (Hixie) [inline packages/flutter/lib/src/services/text_input.dart]: this isn't a JSON object, it's a Map...

File: packages/flutter/lib/src/services/text_input.dart [hunk 3]
```diff
@@ -276,6 +293,8 @@ class TextInputConnection {
 
 TextInputAction _toTextInputAction(String action) {
   switch (action) {
+    case 'TextInputAction.newline':
+      return TextInputAction.newline;
     case 'TextInputAction.done':
       return TextInputAction.done;
```
Comment (Hixie) [inline packages/flutter/lib/src/services/text_input.dart]: you should keep the switch and the enum definition in the same order unless the order matters

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 4]
```diff
@@ -141,11 +141,14 @@ class EditableText extends StatefulWidget {
   /// Creates a basic text input control.
   ///
   /// The [maxLines] property can be set to null to remove the restriction on
-  /// the number of lines. By default, it is 1, meaning this is a single-line
-  /// text field. If it is not null, it must be greater than zero.
+  /// the number of lines. By default, it is one, meaning this is a single-line
+  /// text field. [maxLines] must not be set to zero.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/editable_text.dart]: According to your change here, it's now allowed to be negative, which I doubt is what you intended.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 5]
```diff
@@ -141,11 +141,14 @@ class EditableText extends StatefulWidget {
...
   ///
-  /// The [controller], [focusNode], [style], [cursorColor], and [textAlign]
-  /// arguments must not be null.
+  /// If [maxLines] is not one, then [keyboardType] will be ignored, and the
+  /// keyboard type [TextInputType.multiline] will be used.
```
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I think it'd be cleaner if the default for keyboardType depended on maxLines, but if it was set, it was honoured regardless.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 6]
```diff
@@ -194,7 +199,7 @@ class EditableText extends StatefulWidget {
   /// Defaults to false.
   final bool obscureText;
 
-  /// Whether to enable autocorrection.
+  /// Whether to enable auto-correction.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Wikipedia says "autocorrection" is correct.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 7]
```diff
@@ -419,7 +431,15 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
-      _textInputConnection = TextInput.attach(this, new TextInputConfiguration(inputType: widget.keyboardType, obscureText: widget.obscureText, autocorrect: widget.autocorrect))
+      _textInputConnection = TextInput.attach(this,
+          new TextInputConfiguration(
+              inputType: widget.keyboardType,
+              obscureText: widget.obscureText,
+              autocorrect: widget.autocorrect,
+              inputAction: widget.keyboardType == TextInputType.multiline
+                  ? TextInputAction.newline
+                  : TextInputAction.done
+          ))
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/editable_text.dart]: the formatting is off somehow here. There's two closing parens on the same line, but I don't see any lines with two opening parens.

File: packages/flutter_test/lib/src/test_text_input.dart [hunk 8]
```diff
@@ -28,6 +28,9 @@ class TestTextInput {
 
   int _client = 0;
 
+  /// Arguments supplied to setCient.
```
Comment (Hixie) [inline packages/flutter_test/lib/src/test_text_input.dart]: typo?

File: packages/flutter/lib/src/services/text_input.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/services/text_input.dart]: (no need to fix this unless you want to go on a rampage and fix all of them)

**Comment by Hixie** [comment]
LGTM modulo comments

File: packages/flutter/lib/src/services/text_input.dart [hunk 9]
```diff
@@ -22,6 +22,13 @@ enum TextInputType {
   /// Requests the default platform keyboard.
   text,
 
+  /// Optimize for multi-line textual information.
+  ///
+  /// Requests the default platform keyboard, but accepts newlines when the
+  /// enter key is pressed. This is the input type used for all multi-line text
+  /// fields.
+  multiline,
```
Comment (cbracken) [inline packages/flutter/lib/src/services/text_input.dart]: Would `multilineText` be better, or just more pedantic?

File: packages/flutter/lib/src/services/text_input.dart (hunk 1 above)
Comment (cbracken) [inline packages/flutter/lib/src/services/text_input.dart]: nit: `s/button/key/` also covers physical keyboards.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (cbracken) [reply packages/flutter/lib/src/widgets/editable_text.dart]: +1. I can't imagine much use for a multi-line phone number, but it does seem fair that if someone adds a multi-line text field and request a phone keyboard, that we give them one.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+  final Color cursorColor = const Color.fromARGB(0xFF, 0xFF, 0x00, 0x00);
+
+  testWidgets('has expected defaults', (WidgetTester tester) async {
+    await tester.pumpWidget(new Directionality(
+        textDirection: TextDirection.ltr,
+        child: new EditableText(
+            controller: controller,
+            focusNode: focusNode,
+            style: textStyle,
+            cursorColor: cursorColor)));
```
Comment (cbracken) [inline packages/flutter/test/widgets/editable_text_test.dart]: Here and below: trailing commas.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+    final EditableText editableText =
+        tester.firstWidget(find.byType(EditableText));
+    expect(editableText.maxLines, equals(1));
+    expect(editableText.obscureText, equals(false));
+    expect(editableText.autocorrect, equals(true));
```
Comment (cbracken) [inline packages/flutter/test/widgets/editable_text_test.dart]: isFalse, isTrue.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+                focusNode: focusNode,
+                keyboardType: TextInputType.multiline,
+                style: textStyle,
+                cursorColor: cursorColor))));
+
+    await tester.tap(find.byType(EditableText));
+    await tester.showKeyboard(find.byType(EditableText));
+    controller.text = 'test';
+    await tester.idle();
+    print(tester.testTextInput.editingState.toString());
```
Comment (cbracken) [inline packages/flutter/test/widgets/editable_text_test.dart]: Delete this line.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    expect(tester.testTextInput.editingState['text'], equals('test'));
+    expect(tester.testTextInput.setClientArgs['inputType'],
+        equals('TextInputType.multiline'));
```
Comment (cbracken) [inline packages/flutter/test/widgets/editable_text_test.dart]: Fine to wrap up to prev line. Even though it's a little over 80, it reads a teensy bit better.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+                focusNode: focusNode,
+                maxLines: 3, // Sets multiline keyboard implicitly.
+                style: textStyle,
+                cursorColor: cursorColor))));
+
+    await tester.tap(find.byType(EditableText));
+    await tester.showKeyboard(find.byType(EditableText));
+    controller.text = 'test';
+    await tester.idle();
+    print(tester.testTextInput.editingState.toString());
```
Comment (cbracken) [inline packages/flutter/test/widgets/editable_text_test.dart]: Drop this line.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+                keyboardType: TextInputType.text,
+                maxLines: 3, // Sets multiline keyboard implicitly.
+                style: textStyle,
+                cursorColor: cursorColor))));
+
+    await tester.tap(find.byType(EditableText));
+    await tester.showKeyboard(find.byType(EditableText));
+    controller.text = 'test';
+    await tester.idle();
+    print(tester.testTextInput.editingState.toString());
```
Comment (cbracken) [inline packages/flutter/test/widgets/editable_text_test.dart]: Drop this one too :)

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+                focusNode: focusNode,
+                maxLines: 1, // Sets text keyboard implicitly.
+                style: textStyle,
+                cursorColor: cursorColor))));
+
+    await tester.tap(find.byType(EditableText));
+    await tester.showKeyboard(find.byType(EditableText));
+    controller.text = 'test';
+    await tester.idle();
+    print(tester.testTextInput.editingState.toString());
```
Comment (cbracken) [inline packages/flutter/test/widgets/editable_text_test.dart]: And this one.

**Comment by cbracken** [comment]
Nice!
lgtm
modulo comments too.

**Review by cbracken — APPROVED** [review]

File: packages/flutter_test/lib/src/test_text_input.dart (hunk 8 above)
Comment (gspencergoog) [reply packages/flutter_test/lib/src/test_text_input.dart]: Fixed.

**Pull request merged**